File: backend_nodejs/src/presentation/handler/deleteTodoHandler.ts
import type { NextFunction, Request, Response } from "express";
import z from "zod";
import type { DeleteTodoUsecase } from "@/application_service/usecase/deleteTodoUsecase";
import { RepositoryNotFoundError } from "@/domain/repository/errors";
import {
  BadRequestError,
  NotFoundError,
} from "@/presentation/middleware/errorHandler";

const deleteTodoRequestSchema = z.object({
  params: z.object({
    id: z.uuid({ version: "v7" }),
  }),
});

export class DeleteTodoHandler {
  constructor(private deleteTodoUsecase: DeleteTodoUsecase) {}

  handle = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const parsedReq = deleteTodoRequestSchema.parse(req);
      await this.deleteTodoUsecase.execute({
        id: parsedReq.params.id,
      });
      res.status(204).json({});
    } catch (error) {
      switch (true) {
        case error instanceof z.ZodError: {
          const message = error.issues
            .map((issue) => `${issue.path}: ${issue.message}`)
            .join(", ");
          return next(
            new BadRequestError(`Invalid request parameters: [${message}]`),
          );
        }
        case error instanceof RepositoryNotFoundError:
          return next(new NotFoundError("Todo not found"));
      }
      return next(error);
    }
  };
}

File: backend_python/src/todo_api/presentation/handler/delete_todo_handler.py
from pydantic import BaseModel

from todo_api.application_service.usecase.delete_todo_usecase import (
    DeleteTodoUsecase,
    DeleteTodoUsecaseInput,
)
from todo_api.domain.repository.errors import RepositoryNotFoundError
from todo_api.presentation.middleware.error_handler import NotFoundError
from todo_api.utils.uuid import UUID7


class DeleteTodoResponse(BaseModel):
    pass


class DeleteTodoHandler:
    def __init__(self, delete_todo_usecase: DeleteTodoUsecase) -> None:
        self.delete_todo_usecase = delete_todo_usecase

    def handle(self, id: UUID7) -> DeleteTodoResponse:
        try:
            self.delete_todo_usecase.execute(DeleteTodoUsecaseInput(id=id))
        except RepositoryNotFoundError:
            raise NotFoundError(message="Todo not found")
        return DeleteTodoResponse()
